hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java

```diff
@@ -114,8 +116,26 @@ public TransactionMetadata preHandleDeleteAllowances(TransactionBody txn) {
 
     @Override
     /** {@inheritDoc} */
-    public TransactionMetadata preHandleUpdateAccount(TransactionBody txn) {
-        throw new NotImplementedException();
+    public TransactionMetadata preHandleUpdateAccount(PreHandleContext ctx, TransactionBody txn) {
+        final var op = txn.getCryptoUpdateAccount();
+        final var payer = txn.getTransactionID().getAccountID();
+        final var updateAccountId = op.getAccountIDToUpdate();
+        final var meta = new SigTransactionMetadata(accountStore, txn, payer);
+
+        final var newAccountKeyMustSign =
+                !((CryptoSignatureWaivers) ctx.signatureWaivers())
+                        .isNewKeySignatureWaived(txn, payer);
+        final var targetAccountKeyMustSign =
+                !((CryptoSignatureWaivers) ctx.signatureWaivers())
+                        .isTargetAccountSignatureWaived(txn, payer);
```

Comment:
I'm not liking the cast. I'm also a little confused as to how this works. The `ctx` is created by the application before it calls the service. It needs to know what the right type of signature waiver is that it needs to create. How will it do that?

Reply:
So, is it better to inject `HederaAccountNumbers` into `PreHandleContext`. So the `CryptoSignatureWaiversImpl` can be constructed in the `preHandleUpdateAccount` method ?